### ISSUE TITLE
Reduce frequency of breaking changes to pursuit json format

### ISSUE THREAD

**Issue by hdgarrood** [purescript/purescript#2186]
I think we are having to break the pursuit json format too often, and I would like to try modifying the format in such a way that we can reduce how often we need to break it. I can think of two options:

1. Keep an approximately equivalent approach, but reduce the number of internal compiler data types which appear inside the JSON blob that is sent, such as `Type`, `Constraint`, and `Fixity`. On a conceptual level, it seems to me that the internal representations of these in the compiler ought to be able to change without needing to regenerate everything.
2. Change the whole approach and just copy what Hackage does; namely, generate the HTML client-side and host it on Pursuit.

2) is easier, but makes it significantly easier for people to put nasty stuff on the website. content-security-policy headers should help there (aside: they seem to have disappeared), but it's still unsettling. I want to try 1) first.

**Comment by garyb** [comment]
I think the churn should slow down a bit now anyway (famous last words) as we're done with the operator changes, etc. I guess there will be some changes for constraint kinds, but I'm not sure there's much else planned for 0.10 that should necessitate it?

Anything we can do that makes it less painful to update sounds good to me, although something along the lines of 1 would be my preference also.

**Comment by paf31** [comment]
I don't think we should generate HTML, but rather some data structure which looks like the subset of HTML we need to use. That way, we can restyle later without having to regenerate everything.

**Comment by garyb** [comment]
Markdown? Haha.

**Comment by paf31** [comment]
Another upside of generating HTML or some other markup is that users can generate it for offline reading.

Markdown would work, except for that we need some way to represent syntax highlighting.

**Comment by jplatte** [comment]
Markdown has code blocks. Or, if we don't want every code block to be highlighted as PureScript (which would require some custom code anyway instead of just a Markdown -> HTML converter), any converter that supports GitHub markdown would work (https://help.github.com/articles/creating-and-highlighting-code-blocks/).

**Comment by paf31** [comment]
Code blocks might not be enough, since what constitutes valid code might change over time.

**Comment by hdgarrood** [comment]
Markdown doesn't seem a great fit to me, given that I want some way of indicating links between modules or packages that isn't tied to the url structure that the pursuit web app uses. The current code does have this, and I'd rather keep it that way. Also, we need to be able to produce Hoogle input files from the same data.

**Comment by hdgarrood** [comment]
Regarding churn rate, I think I'd rather approach it from the point of view of thinking about what kinds of changes should cause breaking changes to the format, and what kinds shouldn't. Even if the rate of these kinds of changes is going to slow down now, I think it's still better to decouple these things a little more, just from an architectural perspective. Even changes that should be totally internal to the compiler, e.g. if `Type` was split into a few data types in order to better indicate which constructors can exist at different stages of compilation, can cause breakage at the moment.

**Comment by paf31** [comment]
Perhaps it would be good to create a list of features we want the data format to support?

- Links
- Basic highlighting
- Hoogle index creation

Did I missing anything?

**Comment by hdgarrood** [comment]
I can't immediately think of anything missing from that list. I'm not particularly keen on a sort of ground-up redesign of the format though (if this is where this discussion is heading); at least, not until we've attempted to fix this by attempting option 1) in the original description. I think the format/approach we have right now is relatively good - I am not aware of any problems with it other than this one - and so I'd like to avoid changing it too much if we can. I also think attempting option 1) would give us some useful insight if it did turn out to be unsuccessful.

**Comment by paf31** [comment]
I think the current format is mostly fine too. The only other thing I'd point to is the relatively large generated file size, but that can be addressed by gzipping and configuring Nginx for Pursuit uploads.

**Comment by hdgarrood** [comment]
Ah, yes, good point. Hopefully this would reduce that at the same time; I guess we would end up with a much less rich structure than that of types like `Type` and `Constraint` currently.
